internal/resource/url.go

```diff
@@ -332,6 +341,68 @@ func (f *Fetcher) fetchFromDataURL(u url.URL, dest io.Writer, opts FetchOptions)
 	return f.decompressCopyHashAndVerify(dest, bytes.NewBuffer(url.Data), opts)
 }
 
+// FetchFromAWSMetadata fetches metadata from the `IMDSv2` service if its configured, else
+// it will fallback to the `IMDSv1`. By default, the value of url is an IMDSv1 endpoint
+// i.e. http://169.254.169.254/2019-10-01/user-data.
+func (f *Fetcher) FetchFromAWSMetadata(u url.URL, opts FetchOptions) ([]byte, error) {
+	userdataLatestURL := url.URL{
+		Scheme: "http",
+		Host:   "169.254.169.254",
+		Path:   "latest",
+	}
+	token, err := f.fetchAWSIMDSV2Token(userdataLatestURL.String())
+	if err != nil && err != ErrIMDSV2 {
+		return nil, err
+	}
+	if err != nil && err == ErrIMDSV2 {
+		// Do nothing
+		f.Logger.Info("IMDSv2 service is disabled; fallback to IMDSv1")
+	}
+	if len(token) > 0 && err == nil {
+		opts = FetchOptions{
+			Headers: http.Header{
+				"X-aws-ec2-metadata-token": []string{token},
+			},
+		}
```

Comment:
Is there any reason that we are throwing away the previous contents of `opts` and just putting the metadata token? I would think we would want to essentially copy the previous struct with the addition of the metadata token to whatever pre-existing headers.

Reply:
Here: https://github.com/coreos/ignition/pull/1154/files#diff-bf030bbb0152addedd720e7c13f083876784c0d85007fefb8e6bf75713b5376aR44
the code is passing an empty FetchOptions struct, so I'm doubtful if we're really throwing away the previous contents of `opts`.

Reply:
While we currently might be passing an empty `FetchOptions` struct there's no guarantee that it will always be that way, it's still better to preserve the existing options.